crates/trie/sparse/src/trie.rs

```diff
@@ -0,0 +1,615 @@
+                        SparseNode::new_split_branch(current[common], path[common]),
+                    );
+                    self.nodes.insert(
+                        path.slice(..=common),
+                        SparseNode::new_leaf(path.slice(common + 1..)),
+                    );
+                    self.nodes.insert(
+                        current.slice(..=common),
+                        SparseNode::new_leaf(current.slice(common + 1..)),
+                    );
+
+                    break;
+                }
+                SparseNode::Extension { key, .. } => {
+                    current.extend_from_slice(key);
+                    if !path.starts_with(&current) {
+                        // find the common prefix
+                        let common = current.common_prefix_length(&path);
+
+                        *key = current.slice(current.len() - key.len()..common);
+
+                        // create state mask for new branch node
+                        // NOTE: this might overwrite the current extension node
+                        let branch = SparseNode::new_split_branch(current[common], path[common]);
+                        self.nodes.insert(current.slice(..common), branch);
+
+                        // create new leaf
+                        let new_leaf = SparseNode::new_leaf(path.slice(common + 1..));
+                        self.nodes.insert(path.slice(..=common), new_leaf);
+
+                        // recreate extension to previous child if needed
+                        let key = current.slice(common + 1..);
+                        if !key.is_empty() {
+                            self.nodes.insert(current.slice(..=common), SparseNode::new_ext(key));
+                        }
+
+                        break;
+                    }
+                }
+                SparseNode::Branch { state_mask, .. } => {
+                    let nibble = path[current.len()];
+                    current.push_unchecked(nibble);
+                    if !state_mask.is_bit_set(nibble) {
+                        state_mask.set_bit(nibble);
+                        let new_leaf = SparseNode::new_leaf(path.slice(current.len()..));
+                        self.nodes.insert(current, new_leaf);
+                        break;
+                    }
+                }
+            };
+        }
+    }
+
+    /// Remove leaf node from the trie.
+    pub fn remove_leaf(&mut self, _path: Nibbles) {
+        unimplemented!()
+    }
+
+    /// Return the root of the sparse trie.
+    /// Updates all remaining dirty nodes before calculating the root.
+    pub fn root(&mut self) -> B256 {
+        // take the current prefix set.
+        let mut prefix_set = std::mem::take(&mut self.prefix_set).freeze();
+        let root_rlp = self.rlp_node(Nibbles::default(), &mut prefix_set);
+        if root_rlp.len() == B256::len_bytes() + 1 {
+            B256::from_slice(&root_rlp[1..])
+        } else {
```

Comment:
do you think it worths moving this to a separate function in `alloy-trie`? this happens often and it's not straight away obvious why is it needed

Reply:
https://github.com/alloy-rs/trie/pull/57